update content-management.md (#1310)
demisto/content-docs: content-repo/extra-docs/packs/content-management.md

* content is saved in a bucket in your cloud provider. The examples in the pack are based on Google Cloud Storage, but the pack can be configured to work with other vendors, too.
​
* run a playbook that reads from a configuration file that determines which content to install in your environment. The configuration file contains a list of content to install from your bucket.

In Cortex XSOAR, you can develop and test your content on other machines, before using it in a production environment. You can do this by using [the remote repository feature in the UI](https://docs.paloaltonetworks.com/cortex/cortex-xsoar/6-5/cortex-xsoar-admin/remote-repository/remote-repositories-overview.html) or using the CI/CD process.

> The CI/CD process tested in Cortex XSOAR version 6.0.0 and above, Cortex XSOAR version 8.0.0 supported as well.

Reviewer: ```suggestion
The CI/CD process was developed and verified on Cortex XSOAR version 6.x and 8.x.
```

    
            # ========= UPLOAD TO ARTIFACTS SERVER OPTION =========
            # TODO: Upload to the artifacts server of your choice.
            # Create a file with the service account data
            # use the bucket_upload script to upload your packs to google cloud storage
            python $GITHUB_WORKSPACE/repository/build_related_scripts/bucket_upload.py --service_account $GITHUB_WORKSPACE/service_account.json --packs_directory $NEW_PACKS_FOLDER --branch_name $BRANCH_NAME
            # Delete the service account file
            rm $GITHUB_WORKSPACE/service_account.json
    
    With:

            # ========= UPLOAD DIRECTLY TO YOUR XSOAR MACHINE (WHEN MERGING TO MAIN REPO) =========
            if [ $BRANCH_NAME != master ]; then
              CONFIG_FILE=$(cat xsoar_config.json)
              MARKETPLACE_PACKS_LIST=$(cat $CONFIG_FILE | jq -r '.marketplace_packs')
              # Upload Custom Packs
              demisto-sdk upload --input-config-file /xsoar_config.json
              # Upload MarketPlace Packs
              python3 build_related_scripts/MarketPlaceInstallerFromCICD.py --marketplace-packs-list $MARKETPLACE_PACKS_LIST

> 2. **(Artifact Server only) Configure the the `bucket_upload.py` file.**

Reviewer 2: ```suggestion
2. **(Artifact Server only) Configure the `bucket_upload.py` file.**
```

   
   Update the bucket name, main bucket path and format file before pushing any content from your Git repository. The artifact server is based on Google Cloud Storage.
   
   If using a different storage provider, you need to update the `bucket_upload.py` file with the required artifact server information. Contact Customer Support if you need to change this.
    
3. **Configure the `xsoar_config.json` file.**
    
    The `xsoar_config.json` file defines how custom packs, content packs, lists, and jobs are set up on the production machine. It consists of the following sections:
    
    | Content | Description |
    | ------- | ----------- | 
    |`custom_packs`| Custom content packs, which are installed through the build process.|
    | `marketplace_packs` |Marketplace packs to be installed on the machine.|
    | `lists`| Lists that are created in the machine.|
    | `jobs` | Jobs that are created in the machine.|

    3.1 **Update the ID and URLs for each custom content pack.**
    

>     For a non-artifact server, change the `URL` to the name of the Pack in the following format (using a local URL):

Reviewer 2: ```suggestion
    For a non-artifact server, change the `URL` to the name of the pack in the following format (using a local URL):
```

     `"Url": "Packs\name”` 
    
    **NOTE:** For a non-artifact server, if you want to change the version you need to change it in the branch repository and not in this file.

>     If using an artifact server, update the version in the `URL`. This enables you to have version control. If there was an error in a content pack you can change the version number to an earlier version in your repository. <br/>

Reviewer 2: ```suggestion
    If using an artifact server, update the version in the `URL`. This enables you to have version control. If there was an error in a content pack, you can change the version number to an earlier version in your repository. 
```

    
    ![config_json.png](../../../docs/doc_imgs/reference/XSOAR-CICD/config_json.png)

    3.2 **(Optional) Update Marketplace Content Packs**.
    

>     You can also add any Marketplace Content Packs to update. You have the option to deploy from here with the latest version. If you have a lot of out-of-the-box content, use the `demisto-sdk xsoar-config-file --add-all-marketplace-packs` command which automatically adds all the installed out of the box Content Packs to the configuration file.

Reviewer 2: ```suggestion
    You can also add any Marketplace content packs to update. You have the option to deploy from here with the latest version. If you have a lot of out-of-the-box content, use the `demisto-sdk xsoar-config-file --add-all-marketplace-packs` command which automatically adds all the installed out-of-the-box content packs to the configuration file.
```

    If using an artifact server, open the folder in the artifact server. The hierarchy should appear similar to this:

    ````
    ├── builds
    │   ├── branch-name
    │   │   ├── packs
    │   │   │   ├── pack1
    │   │   │   │   ├── 1.0.0
    │   │   │   │   │   ├── pack1.zip
    │   │   │   ├── pack2
    │   │   │   │   ├── 1.0.1
    │   │   │   │   │   ├── pack2.zip
    │   │   │   ├── 
    │   ├── 
    ├── production
    │   ├── packs
    │   │   ├── pack1
    │   │   │   ├── 1.0.0
    │   │   │   │   │   ├── pack1.zip
    │   │   │   ├── 1.0.1
    │   │   │   │   │   ├── pack1.zip
    │   │   │   ├── 1.1.0
    │   │   │   │   │   ├── pack1.zip
    │   │   │   ├── 
    │   │   ├── pack2
    │   │   │   ├── 1.0.0
    │   │   │   │   │   ├── pack2.zip
    │   │   │   ├── 1.0.1
    │   │   │   │   │   ├── pack2.zip
    │   │   │   ├── 1.0.2
    │   │   │   │   │   ├── pack2.zip
    │   │   │   ├── 
    │   │   ├──

    ````

6. **(Non-artifact server) Install the content to Cortex XSOAR**.
    
    In Cortex XSOAR go to Installed Packs and see your custom content packs. 

7. **(Artifact Server) Add the content to Cortex XSOAR.**

>    You can either use the [XSOAR CI/CD Content Pack](https://xsoar.pan.dev/docs/reference/packs/content-management) or download content manually. If using Google Cloud Services, download the [Google Cloud Storage Content Pack](https://xsoar.pan.dev/docs/reference/integrations/google-cloud-storage) and set up the integration instance. If using another storage application you need to download the appropriate Content Pack such as AWS. 

Reviewer 2: ```suggestion
   You can either use the [XSOAR CI/CD content pack](https://xsoar.pan.dev/docs/reference/packs/content-management) or download content manually. If using Google Cloud Services, download the [Google Cloud Storage content pack](https://xsoar.pan.dev/docs/reference/integrations/google-cloud-storage) and set up the integration instance. If using another storage application you need to download the appropriate content pack such as AWS. 
```

>    - Download the [XSOAR CI/CD Content Pack](https://xsoar.pan.dev/docs/reference/packs/content-management).

Reviewer 2: ```suggestion
   - Download the [XSOAR CI/CD content pack](https://xsoar.pan.dev/docs/reference/packs/content-management).
```

>    The Content Pack includes the Configuration Setup playbook, the configuration setup layout, incident fields, automations, etc. The playbook can run via a job or manually. The playbook fetches the configuration file and loads the contents to the machine. It downloads, and installs the custom content packs and configures lists and jobs if part of the content packs.

Reviewer 2: ```suggestion
   The content pack includes the Configuration Setup playbook, the configuration setup layout, incident fields, automations, etc. The playbook can run via a job or manually. The playbook fetches the configuration file and loads the contents to the machine. It downloads, and installs the custom content packs and configures lists and jobs if part of the content packs.
```

>    The XSOAR CI/CD Content Pack uses either Google Cloud Storage or HTTP requests to fetch the content packs. If running a different storage provider, you need to download the integration (such as AWS - S3). You need to either create or duplicate the **Configuration Setup** incident field and add the provider (such as AWS) as the source. You also need to update the **Configuration Set_up** playbook by adding a task at the same level as Google Cloud Storage.

Reviewer 2: ```suggestion
   The XSOAR CI/CD content pack uses either Google Cloud Storage or HTTP requests to fetch the content packs. If running a different storage provider, you need to download the integration (such as AWS - S3). You need to either create or duplicate the **Configuration Setup** incident field and add the provider (such as AWS) as the source. You also need to update the **Configuration Setup** playbook by adding a task at the same level as Google Cloud Storage.
```

 
   ![cicd_playbook.png](../../../docs/doc_imgs/reference/XSOAR-CICD/cicd_playbook.png)
  

>    - If not using the XSOAR CI/CD Content Pack, install content manually. For example, if using Google Cloud Services integration run the `gcd-download-file` command.

Reviewer 2: ```suggestion
   - If not using the XSOAR CI/CD content pack, install content manually. For example, if using the Google Cloud Services integration run the `gcd-download-file` command.
```

  
8. If you have any detached content which you have downloaded in step 4 in Setup the CI/CD Development Process above, run the following command:

>    `sdk upload --input-config-file -reattach`

Reviewer 2: ```suggestion

   `sdk upload --input-config-file -reattach`
```

   
   The command uploads all the detached content automatically and compares the content to content in Cortex XSOAR. The files that exist in the `SystemPacks` folder remain detached and all other items are attached when uploaded to Cortex XSOAR.    
      

>    **NOTE**: When you upload detached content to Cortex XSOAR it overrides any existing detached content. To keep any changes make a duplicate of the existing detached content in Cortex XSOAR before uploading.  

Reviewer 2: ```suggestion
   **NOTE**: When you upload detached content to Cortex XSOAR it overrides any existing detached content. To keep any changes, make a duplicate of the existing detached content in Cortex XSOAR before uploading.  
```

│   │   │   ├── ...
│   │   ├── ...
```

## Manage a Pull Request

> In order to use this option in Cortex XSOAR platform, download the **XSOAR CI/CD** content pack from Marketplace. (Search for the **Content Management** content pack).

Reviewer 2: ```suggestion
In order to use this option in the Cortex XSOAR platform, download the **XSOAR CI/CD** content pack from Marketplace. (Search for the **Content Management** content pack).
```

In the pull request management there are 2 options: 
- Pull request creation.
- Pull request update.

The platform supports GitHub, GitLab, and Bitbucket.

> The default Git code management is GitHub. If you want to change the default: 

Reviewer 2: ```suggestion
The default Git code management is GitHub. 
```

To change the default values in the **Pull Request Creation - Generic** playbook:
1. In the Cortex XSOAR platform, go to **Playbooks**.
2. Search for and select the **Pull Request Creation - Generic** playbook.
3. Click the *Playbook Triggered* task (located in the beginning of the playbook) and update the values as required:

   | Field | Description | Default |
   | --- | --- | --- |
   | GitIntegration | Git code management. Possible values are github, gitlab, bitbucket. | github |
   | MainBranch | The branch that the pull request will be merged into | master |

>    | ChannelName | The application to send your message to. Possible values are Slack or Teams | --- |

Reviewer 2: ```suggestion
   | ChannelName | The application to send your message to. Possible values are Slack or Teams. | --- |
```

4. Click **Save**.

### Pull Request Creation
To create a new pull request in your repository:
1. In the Cortex XSOAR platform, go to **Incidents**.
2. Click **New Incident**, and a new window will open.
3. In the new window, fill in the following fields:
    - The name of the incident.
    - The type of the incident: Pull Request Creation.
    - The playbook: Pull Request Creation - Generic.
    - In the *CI/CD Pull Request Attachment* field, attach the files to add to the new pull request.

>     - (Optional) - In the *CI/CD Branch* field, write the name of the new branch for the pull request. The name of the branch should be a new one, that wasn't used before. If you don't fill this field, a new name will be generated automatically during the run of the playbook.

Reviewer 2: ```suggestion
    - (Optional) - In the *CI/CD Branch* field, write the name of the new branch for the pull request. The name of the branch should be a new one that wasn't used before. If you don't fill this field, a new name will be generated automatically during the run of the playbook.
```

    - (Optional) - In the *CI/CD Pull Request Comment* field, write a comment for the pull request description.
4. Click **Create New Incident** to create the pull request.
5. If a channel is configured (as described above), at the end of the process a message will be sent with the new pull request details in the configured channel. 

### Pull Request Update
To update an exiting pull request in your repository:
1. In Cortex XSOAR platform, go to **Incidents**.
2. Click **New Incident**, and a new window will open.
3. In the new window,fill in the following fields:
    - The name of the incident.
    - The type of the incident: Pull Request Update.
    - The playbook: Pull Request Creation - Generic.
    - In the *CI/CD Pull Request Attachment* field, attach the files to add or update in the pull request.
    - Choose the pull request source branch from the list in *CI/CD Pull Request Branch* field, *or* write the name of the source branch in the *CI/CD Branch* field. If a name is not given, a new branch name will be generated automatically and a new pull request will be created.
4. Click *Create New Incident* to update the pull request.

> 5. If a channel is configured (as described above), at the end of the process a message will be sent with the updated pull request details in the configured channel.

Reviewer 2: ```suggestion
If a channel is configured (as described above), at the end of the process a message will be sent with the updated pull request details in the configured channel.
```

Custom Pack Installer for the XSOAR CI/CD pack.
- **JobCreator** - Job creator for the XSOAR CI/CD pack.
- **ListCreator** - List creator for the XSOAR CI/CD pack
- **MarketplacePackInstaller** - Marketplace pack installer for the XSOAR CI/CD pack.
​
### Incident Fields
There are 8 incident fields.
- **Branch Name** - The branch from which to fetch the file. Default is the main branch. (Optional)
- **Configuration File Path** - The relative path within the repository to the *xsoar_config.json* file. (Optional)
- **Configuration File Source** - The source of the configuration file.
- **Custom Packs Installed** - Custom packs that were installed.
- **Custom Packs Source** - The source of the custom packs.
- **Jobs Created** - Jobs that were created.
- **Lists Created** - Lists that were created.
- **Marketplace Packs Installed** - Content packs that were installed.
​
### Incident Type
There is 1 incident type - **Configuration Setup**.
​
### Layout
There is 1 layout - **configuration setup**
![layout.png](https://raw.githubusercontent.com/demisto/content/master/Packs/ContentManagement/doc_files/layout.png)
​
### Playbook
There is 1 playbook - **Configuration Setup**

> This playbook will manage the entire configuration process and needs to run through a `Configuration Setup` incident type.<br />

Reviewer 2: ```suggestion

This playbook will manage the entire configuration process and needs to run through a `Configuration Setup` incident type.<br />
```

2. Server configuration keys.

### CI/CD FAQs

1. **We use Dev/Prod.  Do I still need this feature?  What is the difference?**

>     This feature is an alternative for the Dev/Prod environment. Instead of building and maintaining your code on a development Cortex XSOAR platform, you can do so from your own chosen git repository and utilize 3rd party tools like CI/CD infrastructures, build servers, artifact servers and more. This adds to your development and testing process. These multiple powerful tools do not exist with the Cortex XSOAR UI feature like Version control, code review, distributed developing environments, automatic testing, etc.

Reviewer 2: ```suggestion
    This feature is an alternative for the Dev/Prod environment. Instead of building and maintaining your code on a development Cortex XSOAR platform, you can do so from your own chosen Git repository and utilize 3rd party tools like CI/CD infrastructures, build servers, artifact servers and more. This adds to your development and testing process. These multiple powerful tools do not exist with the Cortex XSOAR UI feature like version control, code review, distributed developing environments, automatic testing, etc.
```
